Add monitoring linter

assafad opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Monitoring Linter is a Golang linter designed to ensure that in Kubernetes operator projects, 
monitoring-related practices are implemented within the `pkg/monitoring` directory using [operator-observability](https://github.com/machadovilaca/operator-observability/tree/main) methods. It verifies that all metrics, alerts and recording rules registrations are centralized in this directory.
The use of [Prometheus registration methods](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Registerer) is restricted across the entire project.

Fixes https://issues.redhat.com/browse/CNV-36760.

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

nunnatsa commented on test/monitoringlocationlinter/prometheushandler/handler.go:
```diff
@@ -0,0 +1,59 @@
+
+func (h dotHandler) IsRegisterMetricsCall(call *ast.CallExpr) bool {
+	if ident, ok := call.Fun.(*ast.Ident); ok {
+		if ident.Name == "Register" || ident.Name == "MustRegister" {
```
Maybe add a helper func that get string and return bool, so you won't have to duplicate thid check below.

assafad replied:
No duplication after removing the handler.

nunnatsa commented on test/monitoringlocationlinter/prometheushandler/handler.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// GetPrometheusHandler returns a prometheus handler according to the way prometheus was imported in the file.
+func GetPrometheusHandler(file *ast.File) Handler {
```
Not sure dot import is a real scenario for prometheus, so I think you only need the name handler. I actually think you don't need the handler at all. The only thing you need from it is the supprot the named import, to make sure the call is from this package.

assafad replied:
+1, removed the handler.

nunnatsa reviewed (commented):
Did preliminary quick review. Added some comments

machadovilaca commented on test/monitoringlocationlinter/analyzer.go:
```diff
@@ -0,0 +1,48 @@
+func run(pass *analysis.Pass) (interface{}, error) {
+	for _, file := range pass.Files {
+		if shouldLintFile(pass.Fset.File(file.Pos()).Name()) {
+			prometheusHandler := prometheushandler.GetPrometheusHandler(file)
+			if prometheusHandler == nil { // no prometheus import => no use in prometheus in this file; nothing to do here
+				continue
+			}
+
+			ast.Inspect(file, func(node ast.Node) bool {
+				call, ok := node.(*ast.CallExpr)
+				if ok && prometheusHandler.IsRegisterMetricsCall(call) {
+					pass.Reportf(node.Pos(), "monitoring-location-linter: Prometheus metrics should be registered only under pkg/monitoring.")
+					return true
+				}
+
+				return true
+			})
```
nit: I think you can move this to a function named `lintFile(file)`

assafad replied:
Done

machadovilaca reviewed (commented):
just a small nit, otherwise good progress so far!

assafad commented on test/monitoringlocationlinter/analyzer.go:
```diff
@@ -0,0 +1,83 @@
+// isInMonitoringDir returns true if the file is located under pkg/monitoring directory.
+func isInMonitoringDir(filePath string) bool {
+	return strings.Contains(filePath, "pkg/monitoring")
+}
```
@machadovilaca other directories to exclude?

machadovilaca replied:
actually `prometheus.Register` and `prometheus.MustRegister` should not be used in any directory

we should then you can add a new rule, where `operatormetrics.RegisterMetrics`, `operatorrules.RegisterAlerts` and `operatorrules.RegisterRecordingRules` are those yes, only used in `pkg/monitoring`

these functions are the ones we now use to managed observability resources

assafad replied:
+1, will update.

assafad replied:
Updated.

machadovilaca commented on Makefile:
```diff
@@ -19,3 +19,15 @@ promlinter-build:
 .PHONY: promlinter-push
 promlinter-push:
 	${CONTAINER_RUNTIME} push ${IMG}
+
+.PHONY: monitoringlocationlinter-unit-test
+monitoringlocationlinter-unit-test:
+	cd test/monitoringlocationlinter && go test ./...
+
+.PHONY: monitoringlocationlinter-build
+monitoringlocationlinter-build:
+	cd test/monitoringlocationlinter && go build ./cmd/monitoringlocationlinter
+
+.PHONY: monitoringlocationlinter-test
+monitoringlocationlinter-test: monitoringlocationlinter-build
+	cd test/monitoringlocationlinter && ./tests/e2e.sh
```
not sure is also correct, but what if we created a Makefile in `test/monitoringlocationlinter` instead

assafad replied:
We can, but not sure it will be convenient to have to `cd test/monitoringlocationlinter/ && make monitoringlocationlinter-unit-test` in order to run those targets. What is the benefit of having another Makefile?

machadovilaca replied:
```
.PHONY: monitoringlocationlinter-unit-test
monitoringlocationlinter-unit-test:
  ...
```

the benefit is not having this kind of definitions


and should read:
`cd test/monitoringlocationlinter/ && make unit-test`

assafad replied:
Didn't create another Makefile, but now target names are shorter.

machadovilaca commented on test/monitoringlocationlinter/analyzer.go:
```diff
@@ -0,0 +1,83 @@
+// getPrometheusLocalName returns the name Prometheus was imported with in the file.
+// e.g. import prom "github.com/prometheus/client_golang/prometheus"
+// In case Prometheus is not imported, it returns an empty string.
+func getPrometheusLocalName(file *ast.File) string {
```
I think returning `(string, bool)`  would make more sense semantically

```
prometheusLocalName, isImported := getPrometheusLocalName(file)
if !isImported {
	return
}
```

on the usage, you have a comment explaining the meaning of `prometheusLocalName == ""` (and twice in the implementation here),
with this, it would become more clear imo

assafad replied:
+1, will update.

machadovilaca commented:
although this is a package that executes testing, it is nonetheless a package to be distributed/imported by other projects
therefore I think it should be in `cmd/` `pkg/` dirs

`test/monitoringlocationlinter/analyzer.go` should be `pkg/monitoringlocationlinter/analyzer.go`
and
`test/monitoringlocationlinter/cmd/monitoringlocationlinter/main.go` should be `cmd/monitoringlocationlinter/main.go`

code in `test/` should be used to test the code in the own repo `cmd/`, `pkg/`, `internal/`, etc... dirs

wdyt @nunnatsa?

nunnatsa commented:
> although this is a package that executes testing, it is nonetheless a package to be distributed/imported by other projects therefore I think it should be in `cmd/` `pkg/` dirs
> 
> `test/monitoringlocationlinter/analyzer.go` should be `pkg/monitoringlocationlinter/analyzer.go` and `test/monitoringlocationlinter/cmd/monitoringlocationlinter/main.go` should be `cmd/monitoringlocationlinter/main.go`
> 
> code in `test/` should be used to test the code in the own repo `cmd/`, `pkg/`, `internal/`, etc... dirs
> 
> wdyt @nunnatsa?

It should not be in `tests` for sure. But I would add everything under one directory, e.g.

(and by the way, "monitoringlocationlinter" limit you to only check locations. What if you'll want to add additional rules?)

WDYT about this?
```
monitorinlinter
+---> cmd/
          + main.go
+---> analyzer/
          +---> testdata/
                    ...
          + -- analyzer.go
          + -- analyzer_test.go
```

nunnatsa commented on test/monitoringlocationlinter/analyzer.go:
```diff
@@ -0,0 +1,83 @@
+// isRegisterMetricsCall returns true if a call is a register metrics call.
+func isRegisterMetricsCall(call *ast.CallExpr, prometheusLocalName string) bool {
+	selectorExpr, ok := call.Fun.(*ast.SelectorExpr)
+	if ok && (selectorExpr.Sel.Name == "Register" || selectorExpr.Sel.Name == "MustRegister") {
+		if ident, ok := selectorExpr.X.(*ast.Ident); ok {
+			if ident.Name == prometheusLocalName {
+				return true
+			}
+		}
+	}
```
nit: there is no logic problem here. But for me, reading it is a bit confusing. I would first check if this is a call from the prometheus package, then check the specific function name. 

WDYT about:
```suggestion
	if !ok {
		return false
	}
	
	if ident, ok := selectorExpr.X.(*ast.Ident); !ok || ident.Name == prometheusLocalName {
			return false
	}
	
	switch selectorExpr.Sel.Name {
	case "Register", "MustRegister":
		return true
	}
```

assafad commented:
> > although this is a package that executes testing, it is nonetheless a package to be distributed/imported by other projects therefore I think it should be in `cmd/` `pkg/` dirs
> > `test/monitoringlocationlinter/analyzer.go` should be `pkg/monitoringlocationlinter/analyzer.go` and `test/monitoringlocationlinter/cmd/monitoringlocationlinter/main.go` should be `cmd/monitoringlocationlinter/main.go`
> > code in `test/` should be used to test the code in the own repo `cmd/`, `pkg/`, `internal/`, etc... dirs
> > wdyt @nunnatsa?
> 
> It should not be in `tests` for sure. But I would add everything under one directory, e.g.
> 
> (and by the way, "monitoringlocationlinter" limit you to only check locations. What if you'll want to add additional rules?)
> 
> WDYT about this?
> 
> ```
> monitoringlinter/
> + -- go.mod
> + -- go.sum
> +---> cmd/
>           + main.go
> +---> analyzer/
>           +---> testdata/
>                     ...
>           + -- analyzer.go
>           + -- analyzer_test.go
> ```

@nunnatsa +1 for changing to `monitoringlinter`. Are you suggesting to have its dir under the root dir?

assafad commented:
Hi @nunnatsa, @machadovilaca, can you please review again? 
I updated the PR with the following changes:
- Changed the linter name to `monitoringlinter` and moved its dir to the root dir, as discussed. 
I find the current linter dir structure the most convenient, but let me know if don't agree.
- Following @machadovilaca's comment, to add checks for operator-observability packages, I changed `analyzer.go`, please have a look at it again.
- Updated all required docs and tests following this changes (e.g. added unit tests and files to the testdata).

nunnatsa commented on monitoringlinter/README.md:
````diff
@@ -0,0 +1,82 @@
+
+**Examples**:
+
+- The following code is restricted across all the project, as it is using Prometheus package registration methods directly:
+    ```go
+    package main
+    
+    import "github.com/prometheus/client_golang/prometheus"
+    
+    func main() {
+        prometheus.Register(prometheus.NewGauge(
+            prometheus.GaugeOpts{
+                Name: "gauge_name",
+                Help: "gauge_help",
+            }))
+    
+        prometheus.MustRegister(prometheus.NewCounter(
+            prometheus.CounterOpts{
+                Name: "counter_name",
+                Help: "counter_help",
+            }))
+    }
+    ```
+
+- The following code is allowed only within `pkg/monitoring`, as it is registering metrics, alerts and recording rules, 
````
Not sure you need the whole code here. WDYT about just listing the not-allowed functions?

assafad replied:
Agreed and updated.

nunnatsa commented on monitoringlinter/README.md:
````diff
@@ -0,0 +1,82 @@
+
+- The following code is allowed only within `pkg/monitoring`, as it is registering metrics, alerts and recording rules, 
+using operator-observability package methods:
+    ```go
+    package monitoring
+    
+    import (
+        "github.com/machadovilaca/operator-observability/pkg/operatormetrics"
+        "github.com/machadovilaca/operator-observability/pkg/operatorrules"
+    )
+    
+    func register() {
+        operatormetrics.RegisterMetrics()
+        operatorrules.RegisterAlerts()
+        operatorrules.RegisterRecordingRules()
+    }
````
same

nunnatsa commented on monitoringlinter/README.md:
````diff
@@ -0,0 +1,82 @@
+    ```
+
+- The following code is allowed across all the project, as it is not registering metrics, alerts and recording rules:
+    ```go
+    package main
+    
+    import (
+        "github.com/machadovilaca/operator-observability/pkg/operatormetrics"
+        "github.com/prometheus/client_golang/prometheus"
+    )
+    
+    func main() {
+        prometheus.NewHistogram(prometheus.HistogramOpts{
+            Name: "hist_name",
+            Help: "hist_help",
+        })
+        
+        operatormetrics.ListMetrics()
+    }
+    ```
````
same

nunnatsa commented on monitoringlinter/go.mod:
```diff
@@ -0,0 +1,7 @@
+module github.com/monitoring/monitoringlinter
+
+go 1.22.0
```
please keep it on go1.21 for now. 

It's going to be used in projects that are still using 1.21 and these project won't be able to use this package.

assafad replied:
Done

nunnatsa commented on monitoringlinter/testdata/src/a/testrepo/go.mod:
```diff
@@ -0,0 +1,36 @@
+module testrepo
+
+go 1.22.0
```
same

nunnatsa commented:
/lgtm
/approve
